File: backend-python/com/gxs/bank/service/BeneficiaryService.py
from com.gxs.bank.exception.ResourceNotFoundException import ResourceNotFoundException
from com.gxs.bank.model.Beneficiary import Beneficiary
from com.gxs.bank.repository.BeneficiaryRepository import BeneficiaryRepository
from com.gxs.bank.repository.UserRepository import UserRepository


class BeneficiaryService:
    def __init__(self, db):
        self.db = db
        self.beneficiaryRepository = BeneficiaryRepository(db)
        self.userRepository = UserRepository(db)

    def addBeneficiary(self, userId: str, request: Beneficiary):
        user = self.userRepository.findById(userId)
        if user is None:
            raise ResourceNotFoundException("User not found")
        request.user = user
        request.userId = user.id
        self.beneficiaryRepository.save(request)
        self.db.commit()
        self.db.refresh(request)
        return request

    def getUserBeneficiaries(self, userId: str):
        return self.beneficiaryRepository.findByUserId(userId)

    def deleteBeneficiary(self, beneficiaryId: str, userId: str):
        beneficiary = self.beneficiaryRepository.findById(beneficiaryId)
        if beneficiary is None:
            raise ResourceNotFoundException("Beneficiary not found")
        if beneficiary.userId != userId:
            raise RuntimeError("Unauthorized")
        self.beneficiaryRepository.delete(beneficiary)
        self.db.commit()

File: backend/src/main/java/com/gxs/bank/service/BeneficiaryService.java
package com.gxs.bank.service;

import com.gxs.bank.exception.ResourceNotFoundException;
import com.gxs.bank.model.Beneficiary;
import com.gxs.bank.model.User;
import com.gxs.bank.repository.BeneficiaryRepository;
import com.gxs.bank.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class BeneficiaryService {
    private final BeneficiaryRepository beneficiaryRepository;
    private final UserRepository userRepository;

    public Beneficiary addBeneficiary(UUID userId, Beneficiary request) {
        User user = userRepository.findById(userId).orElseThrow(() -> new ResourceNotFoundException("User not found"));
        request.setUser(user);
        return beneficiaryRepository.save(request);
    }

    public List<Beneficiary> getUserBeneficiaries(UUID userId) {
        return beneficiaryRepository.findByUserId(userId);
    }

    public void deleteBeneficiary(UUID id, UUID userId) {
        Beneficiary beneficiary = beneficiaryRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("Beneficiary not found"));
        if (!beneficiary.getUser().getId().equals(userId)) {
            throw new RuntimeException("Unauthorized");
        }
        beneficiaryRepository.delete(beneficiary);
    }
}
